clients/venice-producer/src/main/java/com/linkedin/venice/producer/online/OnlineVeniceProducer.java

```diff
@@ -42,26 +44,50 @@ public class OnlineVeniceProducer<K, V> extends AbstractVeniceProducer<K, V> {
   private final ICProvider icProvider;
 
   OnlineVeniceProducer(
-      InternalAvroStoreClient<K, V> storeClient,
-      SchemaReader kmeSchemaReader,
+      ClientConfig storeClientConfig,
       VeniceProperties producerConfigs,
       MetricsRepository metricsRepository,
       ICProvider icProvider) {
-    this.storeClient = storeClient;
-    this.storeName = storeClient.getStoreName();
-    this.icProvider = icProvider;
-
     Duration schemaRefreshPeriod;
     if (producerConfigs.containsKey(CLIENT_PRODUCER_SCHEMA_REFRESH_INTERVAL_SECONDS)) {
       schemaRefreshPeriod =
           Duration.ofSeconds(producerConfigs.getLong(CLIENT_PRODUCER_SCHEMA_REFRESH_INTERVAL_SECONDS));
     } else {
-      schemaRefreshPeriod = ClientConfig.DEFAULT_SCHEMA_REFRESH_PERIOD;
+      schemaRefreshPeriod = storeClientConfig.getSchemaRefreshPeriod();
+    }
+
+    AvroGenericStoreClient<K, V> tempStoreClient = ClientFactory.getAndStartAvroClient(storeClientConfig);
+
+    if (!(tempStoreClient instanceof InternalAvroStoreClient)) {
+      throw new IllegalStateException(
+          "Store client must be of type " + InternalAvroStoreClient.class.getCanonicalName() + ". Got "
+              + getClass().getCanonicalName());
     }
 
-    this.schemaReader =
-        new RouterBackedSchemaReader(storeClient, Optional.empty(), Optional.empty(), schemaRefreshPeriod, icProvider);
-    configure(storeName, producerConfigs, metricsRepository, schemaReader, kmeSchemaReader);
+    this.storeClient = (InternalAvroStoreClient<K, V>) tempStoreClient;
+    this.storeName = storeClient.getStoreName();
+    this.icProvider = icProvider;
+
+    ClientConfig clientConfigForSchemaReader =
+        ClientConfig.cloneConfig(storeClientConfig).setSchemaRefreshPeriod(schemaRefreshPeriod);
+
+    this.schemaReader = ClientFactory.getSchemaReader(clientConfigForSchemaReader, icProvider);
+    ClientConfig<KafkaMessageEnvelope> kmeClientConfig = ClientConfig.cloneConfig(storeClientConfig)
+        .setStoreName(AvroProtocolDefinition.KAFKA_MESSAGE_ENVELOPE.getSystemStoreName())
+        .setSpecificValueClass(KafkaMessageEnvelope.class);
+
+    try (SchemaReader kmeSchemaReader = ClientFactory.getSchemaReader(kmeClientConfig, icProvider)) {
+      configure(storeName, producerConfigs, metricsRepository, schemaReader, kmeSchemaReader);
+    } catch (Throwable e) {
+      // When using D2TransportClient, the threads that D2 creates are non-daemon threads which hold up graceful
+      // termination of the "main" thread. This is especially problematic for the shell producer.
+      Utils.closeQuietlyWithErrorLogged(this);
+      if (e instanceof RuntimeException) {
```

Comment:
VeniceException is extended from RTE, so this will hit true to all VeniceException which might be thrown from getSchemaReader.  do you want to cast those?

Reply:
Yes. The idea here is to throw all runtime exceptions directly. Casting VeniceException to RuntimeException doesn't change it's type. Only how the compiler sees it. Ideally, I wanted to do:
```
if (e instanceof RuntimeException) {
    throw e;
}
```

But the compiler doesn't like it as it needs to be explicitly told the type. Anything that is not a `RuntimeException` cannot be thrown unchecked. So, we wrap it in a `VeniceException` to make it a `RuntimeException`